controller/RestTest.php

```diff
@@ -15,78 +15,118 @@
             try {
-                $customParams = [];
-                if ($this->hasRequestParameter(self::REST_DELIVERY_PARAMS)) {
-                    $customParams = $this->getRequestParameter(self::REST_DELIVERY_PARAMS);
-                    $customParams = json_decode(html_entity_decode($customParams), true);
-                }
-                $deliveryClassLabel = '';
-                if ($this->hasRequestParameter(self::REST_DELIVERY_CLASS_LABEL)) {
-                    $deliveryClassLabel = $this->getRequestParameter(self::REST_DELIVERY_CLASS_LABEL);
-                }
-                $task = ImportAndCompile::createTask($importerId, $file, $customParams, $deliveryClassLabel);
+                $importerId = $this->getParameter(self::REST_IMPORTER_ID);
+                $file = HttpHelper::getUploadedFile(self::REST_FILE_NAME);
+                $customParams = $this->getDecodedParameter(self::REST_DELIVERY_PARAMS);
+                $deliveryClassLabels = $this->getDecodedParameter(self::REST_DELIVERY_CLASS_LABELS);
+
+                $task = ImportAndCompile::createTask($importerId, $file, $customParams, $deliveryClassLabels);
             } catch (ImporterNotFound $e) {
-                $this->returnFailure(new \common_exception_NotFound($e->getMessage()));
+                $this->returnFailure(new NotFoundException($e->getMessage()));
             }
-            
+
             $result = ['reference_id' => $task->getId()];
 
             /** @var TaskLogInterface $taskLog */
             $taskLog = $this->getServiceLocator()->get(TaskLogInterface::SERVICE_ID);
-
             $report = $taskLog->getReport($task->getId());
 
-            if (!empty($report)) { //already executed
-                if ($report instanceof \common_report_Report) {
-                    //serialize report to array
-                    $report = json_encode($report);
-                    $report = json_decode($report);
+            if (!empty($report)) {
+                if ($report instanceof Report) {
+                    // Serialize report to array
+                    $report = json_decode(json_encode($report));
                 }
+
                 $result['common_report_Report'] = $report;
             }
 
-            return $this->returnSuccess($result);
+            $this->returnSuccess($result);
         } else {
-            return $this->returnFailure(new \common_exception_BadRequest('Test package file was not given'));
+            $this->returnFailure(new BadRequestException('Test package file was not given'));
+        }
+    }
+
+    /**
+     * @param string $name
+     * @param mixed|null $default
+     *
+     * @return string|null
+     */
+    private function getParameter(string $name, $default = null): ?string
+    {
+        if (!isset($this->body)) {
+            $this->body = $this->getPsrRequest()->getParsedBody();
         }
+
+        return $this->body[$name] ?? $default;
```

Comment:
I don't think that `body` memoization here makes any difference. The only overhead is two methods call, both of which do not perform any computations, they just return a value.

One more important thing to note here, though, is that PSR `ServerRequestInterface` doesn't guarantee the `getParsedBody` type, it can be either an array or an object. In the later case, your code will fail with a fatal `Uncaught Error: Cannot use object as array` error.
```suggestion
        $bodyParameters = $this->getPsrRequest()->getParsedBody();

        if (is_array($bodyParameters) && isset($bodyParameters[$name])) {
            return $bodyParameters[$name];
        }

        if (is_object($bodyParameters) && property_exists($bodyParameters, $name)) {
            return $bodyParameters->$name;
        }

        return $default;
```